xrootd/client/client.go

```diff
@@ -104,182 +94,83 @@ func NewClient(ctx context.Context, address string, username string, opts ...Opt
-				}
-				resp.Err = err
-			} else if header.Status != xrdproto.Ok {
-				resp.Err = header.Error(resp.Data)
-			}
-
-			if err := client.mux.SendData(header.StreamID, resp); err != nil {
-				if ctx.Err() != nil {
-					// something happened to the context.
-					// ignore this error.
-					continue
-				}
-				panic(err)
-				// TODO: should we just ignore responses to unclaimed stream IDs?
-			}
-
-			if header.Status != xrdproto.OkSoFar {
-				client.mux.Unclaim(header.StreamID)
-			}
-		}
-	}
+	defer client.cancel()
+	// TODO: cleanup sessions.
+	return nil
 }
 
 // Send sends the request to the server and stores the response inside the resp.
-func (client *Client) Send(ctx context.Context, resp xrdproto.Response, req xrdproto.Request) error {
-	data, err := client.call(ctx, req)
-	if err != nil {
-		return err
-	}
-
-	return resp.UnmarshalXrd(xrdenc.NewRBuffer(data))
+// If the resp is nil, then no response is stored.
+func (client *Client) Send(ctx context.Context, resp xrdproto.Response, req xrdproto.Request) (string, error) {
+	return client.sendSession(ctx, client.initialSessionID, resp, req)
 }
 
-func (client *Client) send(ctx context.Context, responseChannel mux.DataRecvChan, request []byte) ([]byte, error) {
-	if _, err := client.conn.Write(request); err != nil {
-		return nil, err
+func (client *Client) sendSession(ctx context.Context, sessionID string, resp xrdproto.Response, req xrdproto.Request) (string, error) {
+	address := sessionID
+	client.mu.RLock()
+	session, ok := client.sessions[address]
+	client.mu.RUnlock()
+	if !ok {
+		return "", errors.Errorf("xrootd: session with id = %q was not found", sessionID)
 	}
 
-	var data []byte
-
-	for {
-		select {
-		case resp, more := <-responseChannel:
-			if !more {
-				return data, nil
-			}
-
-			if resp.Err != nil {
-				return nil, resp.Err
-			}
-
-			data = append(data, resp.Data...)
-		case <-ctx.Done():
-			if err := ctx.Err(); err != nil {
-				return nil, err
-			}
-		}
-	}
-	panic("unreachable")
-}
-
-func (client *Client) call(ctx context.Context, req xrdproto.Request) ([]byte, error) {
-	streamID, responseChannel, err := client.mux.Claim()
+	redirection, err := session.Send(ctx, resp, req)
 	if err != nil {
-		return nil, err
+		return address, err
 	}
 
-	var wBuffer xrdenc.WBuffer
-	header := xrdproto.RequestHeader{streamID, req.ReqID()}
-	if err = header.MarshalXrd(&wBuffer); err != nil {
-		return nil, err
-	}
-	if err = req.MarshalXrd(&wBuffer); err != nil {
-		return nil, err
-	}
-	data := wBuffer.Bytes()
-
-	if client.signRequirements.Needed(req) {
-		data, err = client.sign(streamID, req.ReqID(), data)
+	for cnt := 10; redirection != nil && cnt > 0; cnt-- {
+		address = redirection.Host + ":" + strconv.Itoa(redirection.Port)
+		// TODO: pass token to the session.Login()
+		session, err = client.getSession(address)
 		if err != nil {
-			return nil, err
+			return address, err
+		}
+		if filepathRequest, ok := req.(xrdproto.FilepathRequest); ok {
```

Comment:
`s/filepathRequest/fpReq/`
or perhaps just even:
`s/filepathRequest/fp/`

(we see the type of that new variable in the type-assert, so `filepathRequest` is just noise)